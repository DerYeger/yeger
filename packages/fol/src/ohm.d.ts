import { Node } from 'ohm-js'
import { BoundVariable, Term, Formula } from '~/model'

export type BoundVariables = Set<String>

declare module 'ohm-js' {
  interface Node {
    args: {
      boundVariables: BoundVariables
    }
    parseFormula(boundVariables: BoundVariables): Formula
    parseString(): string
    parseTerm(boundVariables: BoundVariables): Term
    parseTermList(boundVariables: BoundVariables): Term[]
    parseVariable(boundVariables: BoundVariables): BoundVariable
  }

  interface Dict {
    parseFormula(boundVariables: BoundVariables): Formula
  }
}
