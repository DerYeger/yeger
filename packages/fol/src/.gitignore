fol.ohm-bundle.d.ts
fol.ohm-bundle.js
