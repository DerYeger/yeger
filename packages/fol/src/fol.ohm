FOL {
  Formula = BiImplicationFormula

  BiImplicationFormula
    = BiImplicationFormula ("<->" | "↔") ImplicationFormula -- biImplication
    | ImplicationFormula

  ImplicationFormula
    = ImplicationFormula ("->" | "→") OrFormula -- implication
    | OrFormula

  OrFormula
    = OrFormula ("||" | "|" | "∨") AndFormula -- or
    | AndFormula

  AndFormula
    = AndFormula ("&&" | "&" | "∧") UnaryFormula -- and
    | UnaryFormula

  UnaryFormula
    = "(" Formula ")" -- paren
    | ("forall" | "∀") BoundVariable "." Formula -- universal
    | ("exists" | "∃") BoundVariable "." Formula -- existential
    | ("!" | "¬") UnaryFormula -- not
    | Relation
    | BooleanLiteral

  RelationName (a relation)
    = ident

  FunctionName (a function)
    = ident

  BoundVariable (a bound variable)
    = ident

  Relation
    = Term "=" Term -- equality
    | RelationName "(" TermList ")" -- regular

  BooleanLiteral
    = True
    | False

  True = "tt" | "⊤"

  False = "ff" | "⊥"

  Term
    = Function
    | ConstantOrBoundVariable

  ConstantOrBoundVariable (a bound variable or a constant)
    = ident

  ident (an identifier)
    = letter+

  Function
    = FunctionName "(" TermList ")"

  TermList
    = NonemptyListOf<Term, ",">
}
