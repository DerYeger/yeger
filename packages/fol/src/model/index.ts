export * from '~/model/formula'
export * from '~/model/model'
export * from '~/model/term'
