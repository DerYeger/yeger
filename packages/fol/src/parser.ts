import type { FOLSemantics } from '~/fol.ohm-bundle'
import grammar from '~/fol.ohm-bundle'
import type { Expression, Formula } from '~/model'
import {
  AndFormula,
  BiImplFormula,
  BinaryFunction,
  BinaryRelation,
  BooleanLiteral,
  BoundVariable,
  Constant,
  EqualityRelation,
  ExistentialQuantorFormula,
  ImplFormula,
  NotFormula,
  OrFormula,
  ParenthesizedFormula,
  UnaryFunction,
  UnaryRelation,
  UniversalQuantorFormula,
} from '~/model'

const semantics: FOLSemantics = grammar
  .createSemantics()
  .addOperation<string>('parseString()', {
    ident(_) {
      return this.sourceString
    },
  })
  .addOperation<BoundVariable>('parseVariable(boundVariables)', {
    BoundVariable(name) {
      return new BoundVariable(name.parseString())
    },
  })
  .addOperation<Expression>('parseExpression(boundVariables)', {
    ConstantOrBoundVariable(name) {
      const parsedName = name.parseString()
      if (this.args.boundVariables.has(parsedName)) {
        return new BoundVariable(parsedName)
      }
      return new Constant(parsedName)
    },
    UnaryFunction(name, _leftParen, inner, _rightParen) {
      return new UnaryFunction(
        name.parseString(),
        inner.parseExpression(this.args.boundVariables)
      )
    },
    BinaryFunction(
      name,
      _leftParen,
      firstArgument,
      _comma,
      secondArgument,
      _rightParen
    ) {
      return new BinaryFunction(
        name.parseString(),
        firstArgument.parseExpression(this.args.boundVariables),
        secondArgument.parseExpression(this.args.boundVariables)
      )
    },
  })
  .addOperation<Formula>('parseFormula(boundVariables)', {
    OrFormula_or(left, _operator, right) {
      return new OrFormula(
        left.parseFormula(this.args.boundVariables),
        right.parseFormula(this.args.boundVariables)
      )
    },
    AndFormula_and(left, _operator, right) {
      return new AndFormula(
        left.parseFormula(this.args.boundVariables),
        right.parseFormula(this.args.boundVariables)
      )
    },
    ImplFormula_impl(left, _operator, right) {
      return new ImplFormula(
        left.parseFormula(this.args.boundVariables),
        right.parseFormula(this.args.boundVariables)
      )
    },
    BiImplFormula_biimpl(left, _operator, right) {
      return new BiImplFormula(
        left.parseFormula(this.args.boundVariables),
        right.parseFormula(this.args.boundVariables)
      )
    },
    UnaryFormula_paren(_leftParen, inner, _rightParen) {
      return new ParenthesizedFormula(
        inner.parseFormula(this.args.boundVariables)
      )
    },
    UnaryFormula_not(_not, inner) {
      return new NotFormula(inner.parseFormula(this.args.boundVariables))
    },
    UnaryFormula_universal(_quantor, variable, _dot, inner) {
      const parsedVariable = variable.parseVariable(this.args.boundVariables)
      const newBoundVariables = new Set([
        ...this.args.boundVariables,
        parsedVariable.name,
      ])
      return new UniversalQuantorFormula(
        parsedVariable,
        inner.parseFormula(newBoundVariables)
      )
    },
    UnaryFormula_existential(_quantor, variable, _dot, inner) {
      const parsedVariable = variable.parseVariable(this.args.boundVariables)
      const newBoundVariables = new Set([
        ...this.args.boundVariables,
        parsedVariable.name,
      ])
      return new ExistentialQuantorFormula(
        parsedVariable,
        inner.parseFormula(newBoundVariables)
      )
    },
    UnaryRelation(name, _leftParen, expression, _rightParen) {
      return new UnaryRelation(
        name.parseString(),
        expression.parseExpression(this.args.boundVariables)
      )
    },
    BinaryRelation(
      name,
      _leftParen,
      firstExpression,
      _comma,
      secondExpression,
      _rightParen
    ) {
      return new BinaryRelation(
        name.parseString(),
        firstExpression.parseExpression(this.args.boundVariables),
        secondExpression.parseExpression(this.args.boundVariables)
      )
    },
    EqualityRelation(firstExpression, _equality, secondExpression) {
      return new EqualityRelation(
        firstExpression.parseExpression(this.args.boundVariables),
        secondExpression.parseExpression(this.args.boundVariables)
      )
    },
    True(_) {
      return BooleanLiteral.True
    },
    False(_) {
      return BooleanLiteral.False
    },
  })

export function parse(formula: string): Formula {
  const matchResult = grammar.match(formula, 'Formula')
  return semantics(matchResult).parseFormula(new Set())
}
