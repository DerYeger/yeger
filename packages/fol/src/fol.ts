import type { Result } from 'resumon'

import grammar from '~/fol.ohm-bundle'
import type { Model, ModelCheckerMode, ModelCheckerTrace } from '~/model'
import { parse } from '~/parser'

export * from '~/model'
export * from '~/validator'
export * from '~/fol.ohm-bundle'
export type { MatchResult } from 'ohm-js'

function evaluate(model: Model, formula: string): Result<boolean, string> {
  return parse(formula).map((parsedFormula) =>
    parsedFormula.evaluate(model, {})
  )
}

function traceEvaluation(
  mode: ModelCheckerMode,
  expected: boolean,
  model: Model,
  formula: string
): Result<ModelCheckerTrace, string> {
  return parse(formula).map((parsedFormula) =>
    parsedFormula.traceEvaluation(mode, expected, model, {})
  )
}

export const FOL = {
  evaluate,
  match: (formula: string) => grammar.match(formula),
  parse,
  traceEvaluation,
} as const
