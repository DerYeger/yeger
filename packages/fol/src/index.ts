import type { Result } from 'resumon'
import { Err, Ok } from 'resumon'

import grammar from '~/fol.ohm-bundle'
import type {
  Formula,
  Model,
  ModelCheckerMode,
  ModelCheckerTrace,
} from '~/model'
import { parse } from '~/parser'

export * from '~/fol.ohm-bundle'
export * from '~/model'
export * from '~/validator'
export type { MatchResult } from 'ohm-js'

function evaluate(model: Model, formula: string): Result<boolean, string> {
  return parse(formula).andThen<boolean>((parsedFormula) => {
    try {
      return new Ok(parsedFormula.evaluate(model, {}))
    } catch (err) {
      return new Err((err as Error).message)
    }
  })
}

function traceEvaluation(
  mode: ModelCheckerMode,
  expected: boolean,
  model: Model,
  formula: string,
): Result<ModelCheckerTrace, string> {
  return parse(formula).andThen<ModelCheckerTrace>((parsedFormula) => {
    try {
      return new Ok(parsedFormula.traceEvaluation(mode, expected, model, {}))
    } catch (err) {
      return new Err((err as Error).message)
    }
  })
}

function trace(
  mode: ModelCheckerMode,
  expected: boolean,
  model: Model,
  formula: Formula,
): Result<ModelCheckerTrace, string> {
  try {
    return new Ok(formula.traceEvaluation(mode, expected, model, {}))
  } catch (err) {
    return new Err((err as Error).message)
  }
}

export const FOL = {
  evaluate,
  match: (formula: string) => grammar.match(formula),
  parse,
  trace,
  traceEvaluation,
} as const
