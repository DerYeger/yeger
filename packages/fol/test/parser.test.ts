import { describe, expect, it } from 'vitest'

import { FOL } from '../src/index'
import { TestData } from './test-utils'

describe('FOL parser', () => {
  it.each(TestData.validFormulas)('can parse "%s"', (formula) => {
    const result = FOL.parse(formula).get()
    expect(result).toMatchSnapshot()
    const outputString = result.toFormattedString()
    expect(FOL.parse(outputString).get().toFormattedString()).toEqual(
      outputString,
    )
    // console.log(result)
  })

  it('can parse operators', () => {
    const result = FOL.parse('ff && tt || ff').get()
    expect(result.toFormattedString()).toMatchInlineSnapshot('"⊥ ∧ ⊤ ∨ ⊥"')
    expect(result).toMatchInlineSnapshot(`
      OrFormula {
        "left": AndFormula {
          "left": BooleanLiteral {
            "name": "⊥",
            "value": false,
          },
          "operator": "∧",
          "right": BooleanLiteral {
            "name": "⊤",
            "value": true,
          },
        },
        "operator": "∨",
        "right": BooleanLiteral {
          "name": "⊥",
          "value": false,
        },
      }
    `)
  })

  describe('can parse', () => {
    it('existential quantors', () => {
      const result = FOL.parse('exists c . A(c,c)').get()
      expect(result.toFormattedString()).toMatchInlineSnapshot('"∃c. A(c,c)"')
      expect(result).toMatchInlineSnapshot(`
        ExistentialQuantorFormula {
          "inner": RelationFormula {
            "name": "A",
            "terms": [
              BoundVariable {
                "name": "c",
              },
              BoundVariable {
                "name": "c",
              },
            ],
          },
          "quantor": "∃",
          "variable": BoundVariable {
            "name": "c",
          },
        }
      `)
    })

    it('universal quantors', () => {
      const result = FOL.parse('forall c . C(c)').get()
      expect(result.toFormattedString()).toMatchInlineSnapshot('"∀c. C(c)"')
      expect(result).toMatchInlineSnapshot(`
        UniversalQuantorFormula {
          "inner": RelationFormula {
            "name": "C",
            "terms": [
              BoundVariable {
                "name": "c",
              },
            ],
          },
          "quantor": "∀",
          "variable": BoundVariable {
            "name": "c",
          },
        }
      `)
    })

    it('relations', () => {
      const result = FOL.parse('A(a,a, a) && B(b)').get()
      expect(result.toFormattedString()).toMatchInlineSnapshot(
        '"A(a,a,a) ∧ B(b)"',
      )
      expect(result).toMatchInlineSnapshot(`
        AndFormula {
          "left": RelationFormula {
            "name": "A",
            "terms": [
              ConstantTerm {
                "name": "a",
              },
              ConstantTerm {
                "name": "a",
              },
              ConstantTerm {
                "name": "a",
              },
            ],
          },
          "operator": "∧",
          "right": RelationFormula {
            "name": "B",
            "terms": [
              ConstantTerm {
                "name": "b",
              },
            ],
          },
        }
      `)
    })

    it('functions', () => {
      const result = FOL.parse('B(f(d))').get()
      expect(result.toFormattedString()).toMatchInlineSnapshot('"B(f(d))"')
      expect(result).toMatchInlineSnapshot(`
        RelationFormula {
          "name": "B",
          "terms": [
            FunctionTerm {
              "argumentTerms": [
                ConstantTerm {
                  "name": "d",
                },
              ],
              "name": "f",
            },
          ],
        }
      `)
    })

    it('long identifiers', () => {
      const result = FOL.parse('MyRelation(myFunction(a), second)').get()
      expect(result.toFormattedString()).toMatchInlineSnapshot(
        '"MyRelation(myFunction(a),second)"',
      )
      expect(result).toMatchInlineSnapshot(`
        RelationFormula {
          "name": "MyRelation",
          "terms": [
            FunctionTerm {
              "argumentTerms": [
                ConstantTerm {
                  "name": "a",
                },
              ],
              "name": "myFunction",
            },
            ConstantTerm {
              "name": "second",
            },
          ],
        }
      `)
    })
  })

  it('can parse parentheses', () => {
    const result = FOL.parse('(ff || tt) && tt').get()
    expect(result.toFormattedString()).toMatchInlineSnapshot('"(⊥ ∨ ⊤) ∧ ⊤"')
  })
})
