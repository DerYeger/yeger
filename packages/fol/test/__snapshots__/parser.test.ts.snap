// Vitest Snapshot v1

exports[`FOL parser > can parse "!(tt && !(ff) && tt)" 1`] = `
NotFormula {
  "inner": ParenthesizedFormula {
    "inner": AndFormula {
      "left": AndFormula {
        "left": BooleanLiteral {
          "name": "⊤",
          "value": true,
        },
        "operator": "∧",
        "right": NotFormula {
          "inner": ParenthesizedFormula {
            "inner": BooleanLiteral {
              "name": "⊥",
              "value": false,
            },
          },
          "operator": "¬",
        },
      },
      "operator": "∧",
      "right": BooleanLiteral {
        "name": "⊤",
        "value": true,
      },
    },
  },
  "operator": "¬",
}
`;

exports[`FOL parser > can parse "!(tt && !ff) && tt" 1`] = `
AndFormula {
  "left": NotFormula {
    "inner": ParenthesizedFormula {
      "inner": AndFormula {
        "left": BooleanLiteral {
          "name": "⊤",
          "value": true,
        },
        "operator": "∧",
        "right": NotFormula {
          "inner": BooleanLiteral {
            "name": "⊥",
            "value": false,
          },
          "operator": "¬",
        },
      },
    },
    "operator": "¬",
  },
  "operator": "∧",
  "right": BooleanLiteral {
    "name": "⊤",
    "value": true,
  },
}
`;

exports[`FOL parser > can parse "!A(x)" 1`] = `
NotFormula {
  "inner": RelationFormula {
    "name": "A",
    "terms": [
      ConstantTerm {
        "name": "x",
      },
    ],
  },
  "operator": "¬",
}
`;

exports[`FOL parser > can parse "!ff" 1`] = `
NotFormula {
  "inner": BooleanLiteral {
    "name": "⊥",
    "value": false,
  },
  "operator": "¬",
}
`;

exports[`FOL parser > can parse "!tt && !(ff && tt)" 1`] = `
AndFormula {
  "left": NotFormula {
    "inner": BooleanLiteral {
      "name": "⊤",
      "value": true,
    },
    "operator": "¬",
  },
  "operator": "∧",
  "right": NotFormula {
    "inner": ParenthesizedFormula {
      "inner": AndFormula {
        "left": BooleanLiteral {
          "name": "⊥",
          "value": false,
        },
        "operator": "∧",
        "right": BooleanLiteral {
          "name": "⊤",
          "value": true,
        },
      },
    },
    "operator": "¬",
  },
}
`;

exports[`FOL parser > can parse "!tt" 1`] = `
NotFormula {
  "inner": BooleanLiteral {
    "name": "⊤",
    "value": true,
  },
  "operator": "¬",
}
`;

exports[`FOL parser > can parse "(forall x. A(x)) && (exists x. B(x))" 1`] = `
AndFormula {
  "left": ParenthesizedFormula {
    "inner": UniversalQuantorFormula {
      "inner": RelationFormula {
        "name": "A",
        "terms": [
          BoundVariable {
            "name": "x",
          },
        ],
      },
      "quantor": "∀",
      "variable": BoundVariable {
        "name": "x",
      },
    },
  },
  "operator": "∧",
  "right": ParenthesizedFormula {
    "inner": ExistentialQuantorFormula {
      "inner": RelationFormula {
        "name": "B",
        "terms": [
          BoundVariable {
            "name": "x",
          },
        ],
      },
      "quantor": "∃",
      "variable": BoundVariable {
        "name": "x",
      },
    },
  },
}
`;

exports[`FOL parser > can parse "(tt || ff) && tt" 1`] = `
AndFormula {
  "left": ParenthesizedFormula {
    "inner": OrFormula {
      "left": BooleanLiteral {
        "name": "⊤",
        "value": true,
      },
      "operator": "∨",
      "right": BooleanLiteral {
        "name": "⊥",
        "value": false,
      },
    },
  },
  "operator": "∧",
  "right": BooleanLiteral {
    "name": "⊤",
    "value": true,
  },
}
`;

exports[`FOL parser > can parse "A(x) & B(x)" 1`] = `
AndFormula {
  "left": RelationFormula {
    "name": "A",
    "terms": [
      ConstantTerm {
        "name": "x",
      },
    ],
  },
  "operator": "∧",
  "right": RelationFormula {
    "name": "B",
    "terms": [
      ConstantTerm {
        "name": "x",
      },
    ],
  },
}
`;

exports[`FOL parser > can parse "A(x) && B(x)" 1`] = `
AndFormula {
  "left": RelationFormula {
    "name": "A",
    "terms": [
      ConstantTerm {
        "name": "x",
      },
    ],
  },
  "operator": "∧",
  "right": RelationFormula {
    "name": "B",
    "terms": [
      ConstantTerm {
        "name": "x",
      },
    ],
  },
}
`;

exports[`FOL parser > can parse "A(x) <-> B(x)" 1`] = `
BiImplicationFormula {
  "left": RelationFormula {
    "name": "A",
    "terms": [
      ConstantTerm {
        "name": "x",
      },
    ],
  },
  "operator": "↔",
  "right": RelationFormula {
    "name": "B",
    "terms": [
      ConstantTerm {
        "name": "x",
      },
    ],
  },
}
`;

exports[`FOL parser > can parse "A(x) | B(x)" 1`] = `
OrFormula {
  "left": RelationFormula {
    "name": "A",
    "terms": [
      ConstantTerm {
        "name": "x",
      },
    ],
  },
  "operator": "∨",
  "right": RelationFormula {
    "name": "B",
    "terms": [
      ConstantTerm {
        "name": "x",
      },
    ],
  },
}
`;

exports[`FOL parser > can parse "A(x) || B(x)" 1`] = `
OrFormula {
  "left": RelationFormula {
    "name": "A",
    "terms": [
      ConstantTerm {
        "name": "x",
      },
    ],
  },
  "operator": "∨",
  "right": RelationFormula {
    "name": "B",
    "terms": [
      ConstantTerm {
        "name": "x",
      },
    ],
  },
}
`;

exports[`FOL parser > can parse "A(x) -> B(x)" 1`] = `
ImplicationFormula {
  "left": RelationFormula {
    "name": "A",
    "terms": [
      ConstantTerm {
        "name": "x",
      },
    ],
  },
  "operator": "→",
  "right": RelationFormula {
    "name": "B",
    "terms": [
      ConstantTerm {
        "name": "x",
      },
    ],
  },
}
`;

exports[`FOL parser > can parse "A(x)" 1`] = `
RelationFormula {
  "name": "A",
  "terms": [
    ConstantTerm {
      "name": "x",
    },
  ],
}
`;

exports[`FOL parser > can parse "A(x,y)" 1`] = `
RelationFormula {
  "name": "A",
  "terms": [
    ConstantTerm {
      "name": "x",
    },
    ConstantTerm {
      "name": "y",
    },
  ],
}
`;

exports[`FOL parser > can parse "Relation(x)" 1`] = `
RelationFormula {
  "name": "Relation",
  "terms": [
    ConstantTerm {
      "name": "x",
    },
  ],
}
`;

exports[`FOL parser > can parse "a = b" 1`] = `
EqualityRelationFormula {
  "firstTerm": ConstantTerm {
    "name": "a",
  },
  "secondTerm": ConstantTerm {
    "name": "b",
  },
}
`;

exports[`FOL parser > can parse "a = f(a,b)" 1`] = `
EqualityRelationFormula {
  "firstTerm": ConstantTerm {
    "name": "a",
  },
  "secondTerm": FunctionTerm {
    "argumentTerms": [
      ConstantTerm {
        "name": "a",
      },
      ConstantTerm {
        "name": "b",
      },
    ],
    "name": "f",
  },
}
`;

exports[`FOL parser > can parse "exists a . tt" 1`] = `
ExistentialQuantorFormula {
  "inner": BooleanLiteral {
    "name": "⊤",
    "value": true,
  },
  "quantor": "∃",
  "variable": BoundVariable {
    "name": "a",
  },
}
`;

exports[`FOL parser > can parse "exists x. A(x)" 1`] = `
ExistentialQuantorFormula {
  "inner": RelationFormula {
    "name": "A",
    "terms": [
      BoundVariable {
        "name": "x",
      },
    ],
  },
  "quantor": "∃",
  "variable": BoundVariable {
    "name": "x",
  },
}
`;

exports[`FOL parser > can parse "f(a) = a" 1`] = `
EqualityRelationFormula {
  "firstTerm": FunctionTerm {
    "argumentTerms": [
      ConstantTerm {
        "name": "a",
      },
    ],
    "name": "f",
  },
  "secondTerm": ConstantTerm {
    "name": "a",
  },
}
`;

exports[`FOL parser > can parse "f(g(x)) = y" 1`] = `
EqualityRelationFormula {
  "firstTerm": FunctionTerm {
    "argumentTerms": [
      FunctionTerm {
        "argumentTerms": [
          ConstantTerm {
            "name": "x",
          },
        ],
        "name": "g",
      },
    ],
    "name": "f",
  },
  "secondTerm": ConstantTerm {
    "name": "y",
  },
}
`;

exports[`FOL parser > can parse "f(x) = y" 1`] = `
EqualityRelationFormula {
  "firstTerm": FunctionTerm {
    "argumentTerms": [
      ConstantTerm {
        "name": "x",
      },
    ],
    "name": "f",
  },
  "secondTerm": ConstantTerm {
    "name": "y",
  },
}
`;

exports[`FOL parser > can parse "ff" 1`] = `
BooleanLiteral {
  "name": "⊥",
  "value": false,
}
`;

exports[`FOL parser > can parse "forall a. tt" 1`] = `
UniversalQuantorFormula {
  "inner": BooleanLiteral {
    "name": "⊤",
    "value": true,
  },
  "quantor": "∀",
  "variable": BoundVariable {
    "name": "a",
  },
}
`;

exports[`FOL parser > can parse "forall x. A(x)" 1`] = `
UniversalQuantorFormula {
  "inner": RelationFormula {
    "name": "A",
    "terms": [
      BoundVariable {
        "name": "x",
      },
    ],
  },
  "quantor": "∀",
  "variable": BoundVariable {
    "name": "x",
  },
}
`;

exports[`FOL parser > can parse "forall x. exists y. forall z. f(x) = y && R(y,z) && R(z,y) || (f(x) = x -> f(z) = a <-> ff)" 1`] = `
UniversalQuantorFormula {
  "inner": ExistentialQuantorFormula {
    "inner": UniversalQuantorFormula {
      "inner": OrFormula {
        "left": AndFormula {
          "left": AndFormula {
            "left": EqualityRelationFormula {
              "firstTerm": FunctionTerm {
                "argumentTerms": [
                  BoundVariable {
                    "name": "x",
                  },
                ],
                "name": "f",
              },
              "secondTerm": BoundVariable {
                "name": "y",
              },
            },
            "operator": "∧",
            "right": RelationFormula {
              "name": "R",
              "terms": [
                BoundVariable {
                  "name": "y",
                },
                BoundVariable {
                  "name": "z",
                },
              ],
            },
          },
          "operator": "∧",
          "right": RelationFormula {
            "name": "R",
            "terms": [
              BoundVariable {
                "name": "z",
              },
              BoundVariable {
                "name": "y",
              },
            ],
          },
        },
        "operator": "∨",
        "right": ParenthesizedFormula {
          "inner": BiImplicationFormula {
            "left": ImplicationFormula {
              "left": EqualityRelationFormula {
                "firstTerm": FunctionTerm {
                  "argumentTerms": [
                    BoundVariable {
                      "name": "x",
                    },
                  ],
                  "name": "f",
                },
                "secondTerm": BoundVariable {
                  "name": "x",
                },
              },
              "operator": "→",
              "right": EqualityRelationFormula {
                "firstTerm": FunctionTerm {
                  "argumentTerms": [
                    BoundVariable {
                      "name": "z",
                    },
                  ],
                  "name": "f",
                },
                "secondTerm": ConstantTerm {
                  "name": "a",
                },
              },
            },
            "operator": "↔",
            "right": BooleanLiteral {
              "name": "⊥",
              "value": false,
            },
          },
        },
      },
      "quantor": "∀",
      "variable": BoundVariable {
        "name": "z",
      },
    },
    "quantor": "∃",
    "variable": BoundVariable {
      "name": "y",
    },
  },
  "quantor": "∀",
  "variable": BoundVariable {
    "name": "x",
  },
}
`;

exports[`FOL parser > can parse "forall x. forall y. A(x, y)" 1`] = `
UniversalQuantorFormula {
  "inner": UniversalQuantorFormula {
    "inner": RelationFormula {
      "name": "A",
      "terms": [
        BoundVariable {
          "name": "x",
        },
        BoundVariable {
          "name": "y",
        },
      ],
    },
    "quantor": "∀",
    "variable": BoundVariable {
      "name": "y",
    },
  },
  "quantor": "∀",
  "variable": BoundVariable {
    "name": "x",
  },
}
`;

exports[`FOL parser > can parse "forall x. forall y. h(x,y) = x" 1`] = `
UniversalQuantorFormula {
  "inner": UniversalQuantorFormula {
    "inner": EqualityRelationFormula {
      "firstTerm": FunctionTerm {
        "argumentTerms": [
          BoundVariable {
            "name": "x",
          },
          BoundVariable {
            "name": "y",
          },
        ],
        "name": "h",
      },
      "secondTerm": BoundVariable {
        "name": "x",
      },
    },
    "quantor": "∀",
    "variable": BoundVariable {
      "name": "y",
    },
  },
  "quantor": "∀",
  "variable": BoundVariable {
    "name": "x",
  },
}
`;

exports[`FOL parser > can parse "myConstant = myConstant" 1`] = `
EqualityRelationFormula {
  "firstTerm": ConstantTerm {
    "name": "myConstant",
  },
  "secondTerm": ConstantTerm {
    "name": "myConstant",
  },
}
`;

exports[`FOL parser > can parse "myRelation(MyFunction(someConstant, AnotherConstant))" 1`] = `
RelationFormula {
  "name": "myRelation",
  "terms": [
    FunctionTerm {
      "argumentTerms": [
        ConstantTerm {
          "name": "someConstant",
        },
        ConstantTerm {
          "name": "AnotherConstant",
        },
      ],
      "name": "MyFunction",
    },
  ],
}
`;

exports[`FOL parser > can parse "tt && tt" 1`] = `
AndFormula {
  "left": BooleanLiteral {
    "name": "⊤",
    "value": true,
  },
  "operator": "∧",
  "right": BooleanLiteral {
    "name": "⊤",
    "value": true,
  },
}
`;

exports[`FOL parser > can parse "tt <-> tt" 1`] = `
BiImplicationFormula {
  "left": BooleanLiteral {
    "name": "⊤",
    "value": true,
  },
  "operator": "↔",
  "right": BooleanLiteral {
    "name": "⊤",
    "value": true,
  },
}
`;

exports[`FOL parser > can parse "tt || ff && tt" 1`] = `
OrFormula {
  "left": BooleanLiteral {
    "name": "⊤",
    "value": true,
  },
  "operator": "∨",
  "right": AndFormula {
    "left": BooleanLiteral {
      "name": "⊥",
      "value": false,
    },
    "operator": "∧",
    "right": BooleanLiteral {
      "name": "⊤",
      "value": true,
    },
  },
}
`;

exports[`FOL parser > can parse "tt || tt" 1`] = `
OrFormula {
  "left": BooleanLiteral {
    "name": "⊤",
    "value": true,
  },
  "operator": "∨",
  "right": BooleanLiteral {
    "name": "⊤",
    "value": true,
  },
}
`;

exports[`FOL parser > can parse "tt -> tt" 1`] = `
ImplicationFormula {
  "left": BooleanLiteral {
    "name": "⊤",
    "value": true,
  },
  "operator": "→",
  "right": BooleanLiteral {
    "name": "⊤",
    "value": true,
  },
}
`;

exports[`FOL parser > can parse "tt" 1`] = `
BooleanLiteral {
  "name": "⊤",
  "value": true,
}
`;
